### 마이크로 평균(Micro-average), 매크로 평균(Macro-average) 이란 무엇인가?
 https://unlimitedpower.tistory.com/entry/IR-마이크로-평균Micro-average-매크로-평균Macro-average-이란-무엇인가 [오늘도 공장은 돌아간다.]

  Macro-average, Micro-average 이 둘의 값은 매우 차이가 나게 되면 특정 반(카테고리)에 데이터가 편중된다.

### weighted avg 가중 산술 평균
https://ko.wikipedia.org/wiki/%EA%B0%80%EC%A4%91_%EC%82%B0%EC%88%A0_%ED%8F%89%EA%B7%A0

  
### 분류성능평가지표 - Precision(정밀도), Recall(재현율) and Accuracy(정확도)
https://sumniya.tistory.com/26


### 다중 클래스에 대한 Confusion Matri한 설명
https://towardsdatascience.com/confusion-matrix-for-your-multi-class-machine-learning-model-ff9aa3bf7826

```
------------------------------------------------------------------------------------------------------
TP, FN, FP, TN의 수치로 계산되는 성능 지표 중 대표적으로 쓰이는 것은 정밀도(Precision), 재현율(Recall, Sensitivity), F1 스코어(f1 score)이라고 한다.


```
   
<span style="font-size:125%;">
    Precision(정밀도): 양성에 대하여 거짓,참 여부를 예측할때 진짜로 양성인 경우 <br>    
    Recall(재현율)   : 실제값이 양성을 의미하는 것중 예측을 양성이 사실이라고 직접적으로 한 경우<br>    
    <br>
    F1 스코어(f1 score) : 정밀도와 재현율의 값을 활용한 지표(정밀도값과 재현율값의 조화평균)<br>
    accurancy(정확도) : Accuracy는 TP,TF ,FP FT 모든 지표를 활용한 경우<br><br><br>
</span>    


<span style=" font: bold 1.7em/2em Georgia ;">    
    
$precision  =\frac{TP(실제양성,양성예측(참))}{{TP(실제양성,양성예측(참))}+{FP(실제음성,양성예측(거짓))}}$  



$Recall  =\frac{TP(실제양성,양성예측(참))}{{TP(실제양성,양성예측(참))}+{FN(실제양성,음성예측(거짓))}}$ 
<br><br>
<font size=5>
$f1 score=\frac{2}{{\frac{1}{정밀도}}+{\frac{1}{재현율}}}$<br><br>
$accurancy=\frac{{TP}+{TN}}{TP+TN+FP+FN}$
    
    
</font>
</span> 




# 손글씨

In [5]:
# 필요한 모듈 import하기
from sklearn.datasets import load_digits # 사이킷런 토이데이터셋(손글씨)
from sklearn.model_selection import train_test_split # 분리를 위한 패키지
from sklearn.metrics import classification_report # 정밀도, 재현율, F1 score 제공하는 패키지

#데이터 준비
digits = load_digits()
digits.keys() #digits.key() 어떤 정보들이 담겼을지, keys() 라는 메서드

dict_keys(['data', 'target', 'target_names', 'images', 'DESCR'])

In [6]:
#Feature Data 지정하기
digits_data = digits.data
digits_data.shape # 1797개의 데이터와 64개 특징

(1797, 64)

In [7]:
#Label Data 지정하기
digits_label = digits.target
#Target Names 출력해 보기
digits.target_names # 라벨의 이름을 확인한다.



array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [8]:
#데이터 Describe 해 보기
print(digits.DESCR)

.. _digits_dataset:

Optical recognition of handwritten digits dataset
--------------------------------------------------

**Data Set Characteristics:**

    :Number of Instances: 5620
    :Number of Attributes: 64
    :Attribute Information: 8x8 image of integer pixels in the range 0..16.
    :Missing Attribute Values: None
    :Creator: E. Alpaydin (alpaydin '@' boun.edu.tr)
    :Date: July; 1998

This is a copy of the test set of the UCI ML hand-written digits datasets
https://archive.ics.uci.edu/ml/datasets/Optical+Recognition+of+Handwritten+Digits

The data set contains images of hand-written digits: 10 classes where
each class refers to a digit.

Preprocessing programs made available by NIST were used to extract
normalized bitmaps of handwritten digits from a preprinted form. From a
total of 43 people, 30 contributed to the training set and different 13
to the test set. 32x32 bitmaps are divided into nonoverlapping blocks of
4x4 and the number of on pixels are counted in each blo

In [14]:
#train, test 데이터 분리

X_train, X_test, y_train, y_test = train_test_split(digits_data,    # 데이터 feature 
                                                    digits_label,   # 데이터 label   0~9
                                                    test_size=0.2,# 전체 중 20%만 테스트 데이터로 사용하겠다.
                                                    random_state=7)
print('X_train 개수: ', len(X_train), ', X_test 개수: ', len(X_test))

print("y_test shape: {}".format(y_test.shape) )
print("y_test shape: {}".format(y_test))

X_train 개수:  1437 , X_test 개수:  360
y_test shape: (360,)
y_test shape: [6 0 5 9 2 9 0 4 1 0 1 8 2 5 2 8 1 8 9 1 0 2 0 4 5 3 3 0 0 4 1 4 4 4 6 1 4
 0 6 6 0 9 3 6 6 2 0 1 9 6 2 8 9 9 0 2 0 8 4 6 8 5 8 7 8 7 7 4 1 4 5 5 4 6
 2 0 1 3 7 5 8 2 4 4 2 5 1 9 3 7 6 3 3 5 6 2 1 0 1 9 4 1 1 3 1 6 9 0 3 7 6
 9 3 8 0 8 3 8 8 6 3 7 3 9 0 3 0 9 8 1 2 2 3 6 9 4 0 5 4 2 9 1 0 2 5 0 2 2
 7 4 6 9 8 2 6 0 4 4 8 5 0 2 4 6 8 2 3 7 2 9 0 3 5 9 1 6 8 7 5 3 0 4 2 1 3
 3 6 0 2 8 4 1 4 7 5 7 6 6 8 1 0 6 8 7 1 1 9 8 5 5 3 6 8 1 2 0 7 5 3 0 8 2
 0 4 0 9 4 8 4 7 9 7 3 6 2 5 1 5 9 2 9 9 8 2 1 6 7 1 7 5 7 8 9 5 7 4 3 7 8
 8 2 8 9 5 3 2 8 0 4 2 1 0 8 4 1 7 1 4 7 7 1 8 3 8 4 3 5 9 4 4 8 1 8 7 2 3
 1 1 1 0 2 8 0 7 4 0 1 0 2 3 7 9 8 5 8 2 2 6 5 0 8 9 8 9 0 0 9 7 4 1 2 6 7
 3 7 4 0 2 1 7 2 5 7 2 3 5 7 1 4 1 3 3 8 8 1 0 1 9 3 0]


이미지 데이터 1797개에서 1437개의 훈련데이터, 360개의 테스트데이터

# 다양한 모델로 학습시켜보기

In [215]:
# (1) 필요한 모듈 import
from sklearn.tree import DecisionTreeClassifier # 의사결정트리 패키지
from sklearn.ensemble import RandomForestClassifier # RandomForest 모델
from sklearn import svm # svm모델
from sklearn.linear_model import SGDClassifier #SGD모델
from sklearn.linear_model import LogisticRegression #로지스틱 회귀 모델 
from sklearn.metrics import classification_report  # 정확도(precision,recall,f1-score)

### Decision Tree 사용해 보기

In [216]:
# 모델 학습 및 예측
decision_tree = DecisionTreeClassifier(random_state=32)
decision_tree.fit(X_train, y_train)
y_pred = decision_tree.predict(X_test)
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      0.98      0.99        43
           1       0.81      0.81      0.81        42
           2       0.79      0.82      0.80        40
           3       0.79      0.91      0.85        34
           4       0.83      0.95      0.89        37
           5       0.90      0.96      0.93        28
           6       0.84      0.93      0.88        28
           7       0.96      0.82      0.89        33
           8       0.88      0.65      0.75        43
           9       0.78      0.78      0.78        32

    accuracy                           0.86       360
   macro avg       0.86      0.86      0.86       360
weighted avg       0.86      0.86      0.85       360



In [217]:
from sklearn.metrics import confusion_matrix

confusion_matrix(y_test, y_pred)

array([[42,  0,  0,  1,  0,  0,  0,  0,  0,  0],
       [ 0, 34,  3,  1,  0,  1,  1,  0,  0,  2],
       [ 0,  0, 33,  2,  0,  0,  1,  1,  2,  1],
       [ 0,  1,  0, 31,  0,  0,  0,  0,  1,  1],
       [ 0,  0,  1,  0, 35,  0,  0,  0,  1,  0],
       [ 0,  1,  0,  0,  0, 27,  0,  0,  0,  0],
       [ 0,  0,  0,  0,  2,  0, 26,  0,  0,  0],
       [ 0,  0,  0,  1,  2,  1,  0, 27,  0,  2],
       [ 0,  5,  4,  1,  1,  0,  3,  0, 28,  1],
       [ 0,  1,  1,  2,  2,  1,  0,  0,  0, 25]])

### Random Forest 사용해 보기

In [218]:
random_forest = RandomForestClassifier(random_state=32)
random_forest.fit(X_train, y_train)
y_pred = random_forest.predict(X_test)
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      0.98      0.99        43
           1       0.93      1.00      0.97        42
           2       1.00      1.00      1.00        40
           3       1.00      1.00      1.00        34
           4       0.93      1.00      0.96        37
           5       0.90      0.96      0.93        28
           6       1.00      0.96      0.98        28
           7       0.94      0.97      0.96        33
           8       1.00      0.84      0.91        43
           9       0.94      0.94      0.94        32

    accuracy                           0.96       360
   macro avg       0.96      0.96      0.96       360
weighted avg       0.97      0.96      0.96       360



In [219]:
from sklearn.metrics import confusion_matrix

confusion_matrix(y_test, y_pred)

array([[42,  0,  0,  0,  1,  0,  0,  0,  0,  0],
       [ 0, 42,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 0,  0, 40,  0,  0,  0,  0,  0,  0,  0],
       [ 0,  0,  0, 34,  0,  0,  0,  0,  0,  0],
       [ 0,  0,  0,  0, 37,  0,  0,  0,  0,  0],
       [ 0,  0,  0,  0,  0, 27,  0,  0,  0,  1],
       [ 0,  0,  0,  0,  1,  0, 27,  0,  0,  0],
       [ 0,  0,  0,  0,  0,  0,  0, 32,  0,  1],
       [ 0,  3,  0,  0,  1,  1,  0,  2, 36,  0],
       [ 0,  0,  0,  0,  0,  2,  0,  0,  0, 30]])

### SVM 사용해 보기

In [220]:
svm_model = svm.SVC() # SVC모델 가져오기



svm_model.fit(X_train, y_train) # 모델 훈련 시키기
y_pred = svm_model.predict(X_test) # 모델 테스트하기
print(classification_report(y_test, y_pred)) # 정확도 확인

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        43
           1       0.95      1.00      0.98        42
           2       1.00      1.00      1.00        40
           3       1.00      1.00      1.00        34
           4       1.00      1.00      1.00        37
           5       0.93      1.00      0.97        28
           6       1.00      1.00      1.00        28
           7       1.00      1.00      1.00        33
           8       1.00      0.93      0.96        43
           9       1.00      0.97      0.98        32

    accuracy                           0.99       360
   macro avg       0.99      0.99      0.99       360
weighted avg       0.99      0.99      0.99       360



In [221]:
from sklearn.metrics import confusion_matrix

confusion_matrix(y_test, y_pred)

array([[43,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 0, 42,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 0,  0, 40,  0,  0,  0,  0,  0,  0,  0],
       [ 0,  0,  0, 34,  0,  0,  0,  0,  0,  0],
       [ 0,  0,  0,  0, 37,  0,  0,  0,  0,  0],
       [ 0,  0,  0,  0,  0, 28,  0,  0,  0,  0],
       [ 0,  0,  0,  0,  0,  0, 28,  0,  0,  0],
       [ 0,  0,  0,  0,  0,  0,  0, 33,  0,  0],
       [ 0,  2,  0,  0,  0,  1,  0,  0, 40,  0],
       [ 0,  0,  0,  0,  0,  1,  0,  0,  0, 31]])

### SGD Classifier 사용해 보기<br>

In [222]:
sgd_model = SGDClassifier() #sgd_model 가져오기
sgd_model.fit(X_train, y_train) # 모델 훈련 시키기
y_pred = sgd_model.predict(X_test) # 모델 테스트기기
print(classification_report(y_test, y_pred)) # 정확도 확인

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        43
           1       0.86      0.86      0.86        42
           2       0.98      1.00      0.99        40
           3       0.93      0.79      0.86        34
           4       1.00      0.97      0.99        37
           5       0.61      1.00      0.76        28
           6       0.96      0.93      0.95        28
           7       0.97      0.97      0.97        33
           8       0.90      0.81      0.85        43
           9       0.96      0.72      0.82        32

    accuracy                           0.91       360
   macro avg       0.92      0.91      0.90       360
weighted avg       0.92      0.91      0.91       360



In [223]:
from sklearn.metrics import confusion_matrix

confusion_matrix(y_test, y_pred)

array([[43,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 0, 36,  0,  1,  0,  3,  0,  0,  1,  1],
       [ 0,  0, 40,  0,  0,  0,  0,  0,  0,  0],
       [ 0,  0,  0, 27,  0,  6,  0,  1,  0,  0],
       [ 0,  0,  0,  0, 36,  0,  1,  0,  0,  0],
       [ 0,  0,  0,  0,  0, 28,  0,  0,  0,  0],
       [ 0,  1,  0,  0,  0,  0, 26,  0,  1,  0],
       [ 0,  0,  0,  1,  0,  0,  0, 32,  0,  0],
       [ 0,  4,  1,  0,  0,  3,  0,  0, 35,  0],
       [ 0,  1,  0,  0,  0,  6,  0,  0,  2, 23]])

### Logistic Regression 사용해 보기

In [224]:
logistic_model = LogisticRegression(random_state = 32, max_iter = 2000)#logistic_model 가져오기
#max_iter 해를 찾아가는 데 있어서 반복횟수를 제한한다. 무한루프방지

logistic_model.fit(X_train, y_train) # 모델 훈련 시키기
y_pred = logistic_model.predict(X_test) # 모델 테스트기기
print(classification_report(y_test, y_pred)) # 정확도 확인

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        43
           1       0.95      0.95      0.95        42
           2       0.98      1.00      0.99        40
           3       0.94      0.97      0.96        34
           4       1.00      1.00      1.00        37
           5       0.79      0.96      0.87        28
           6       1.00      0.96      0.98        28
           7       0.97      0.97      0.97        33
           8       0.92      0.84      0.88        43
           9       0.97      0.88      0.92        32

    accuracy                           0.95       360
   macro avg       0.95      0.95      0.95       360
weighted avg       0.96      0.95      0.95       360



### 오류내용  
Increase the number of iterations (max_iter) or scale the data as shown in:   
    https://scikit-learn.org/stable/modules/preprocessing.html   
Please also refer to the documentation for alternative solver options:   
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression   
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)  영
  
  ## 변경내용
  logistic_model = LogisticRegression(random_state = 32, max_iter = 2000)

max_iter 해를 찾아가는 데 있어서 반복횟수를 제한한다. 무한루프방지  
max_iter인자를 추가 하지 않았을때 경고 뜸  
https://blog.naver.com/gdpresent/221703566189 참조

In [225]:
from sklearn.metrics import confusion_matrix

confusion_matrix(y_test, y_pred)

array([[43,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 0, 40,  0,  0,  0,  0,  0,  0,  1,  1],
       [ 0,  0, 40,  0,  0,  0,  0,  0,  0,  0],
       [ 0,  0,  0, 33,  0,  0,  0,  1,  0,  0],
       [ 0,  0,  0,  0, 37,  0,  0,  0,  0,  0],
       [ 0,  0,  0,  0,  0, 27,  0,  0,  1,  0],
       [ 0,  0,  0,  0,  0,  0, 27,  0,  1,  0],
       [ 0,  0,  0,  1,  0,  0,  0, 32,  0,  0],
       [ 0,  2,  1,  0,  0,  4,  0,  0, 36,  0],
       [ 0,  0,  0,  1,  0,  3,  0,  0,  0, 28]])


<div style="font-size:115%; background-color:#eafbee;" >

각 모델에 대한 오차행렬을 봤을때 비중이 거의 TP와 TN로 구성되어 있다.(페이지 상단의 다중클래스오차행렬 사이트참조)  
    정밀도 판단에 사용되는 FP, 재현율에 사용되는 FP는 0에 수렴한다.즉 정밀도수식과 재현율수식이 같아진다. 즉 손글씨에서는 정밀도,재현율, f-1 score 는 같은 수식이라고 생각된다. 성능 평가 지표로 어떤것이 적절하다고 하면 f-1 score로 정하겠다.( 이유는 정밀도와 재현율 두 지표를 활용한 값이기 때문에)  
또한 support수를 확인하면 각 클래스에 대하여 비슷한 숫자들로 잘 분배가 되어 있기에 accuracy도 평가 지표로 활용되기에 적절하다고 생각된다.

5개의 모델의 성능지표를 확인해보면 SVM모델은 모든 면에서 높은 수치를 가지고 있다. 즉 손글씨를 판단할때 SVM모델이 적절하다고 생각된다.

</div>




# 와인 분류하기

In [259]:
# 필요한 모듈 import하기
from sklearn.datasets import load_wine # 사이킷런 토이데이터셋(와인)
from sklearn.model_selection import train_test_split # 분리를 위한 패키지
from sklearn.metrics import classification_report # 정밀도, 재현율, F1 score 제공하는 패키지

#데이터 준비
wine = load_wine()
wine.keys() #load_wine() 어떤 정보들이 담겼을지, keys() 라는 메서드

dict_keys(['data', 'target', 'target_names', 'DESCR', 'feature_names'])

In [260]:
#Feature Data 지정하기
wine_data = wine.data
wine_data.shape # 178개의 데이터와 13개 특징

(178, 13)

In [261]:
#Label Data 지정하기
wine_label = wine.target
#Target Names 출력해 보기
wine.target_names # 라벨의 이름을 확인한다.

array(['class_0', 'class_1', 'class_2'], dtype='<U7')

In [262]:
#데이터 Describe 해 보기
print(wine.DESCR)

.. _wine_dataset:

Wine recognition dataset
------------------------

**Data Set Characteristics:**

    :Number of Instances: 178 (50 in each of three classes)
    :Number of Attributes: 13 numeric, predictive attributes and the class
    :Attribute Information:
 		- Alcohol
 		- Malic acid
 		- Ash
		- Alcalinity of ash  
 		- Magnesium
		- Total phenols
 		- Flavanoids
 		- Nonflavanoid phenols
 		- Proanthocyanins
		- Color intensity
 		- Hue
 		- OD280/OD315 of diluted wines
 		- Proline

    - class:
            - class_0
            - class_1
            - class_2
		
    :Summary Statistics:
    
    ============================= ==== ===== ======= =====
                                   Min   Max   Mean     SD
    ============================= ==== ===== ======= =====
    Alcohol:                      11.0  14.8    13.0   0.8
    Malic Acid:                   0.74  5.80    2.34  1.12
    Ash:                          1.36  3.23    2.36  0.27
    Alcalinity of Ash:            1

In [263]:
#train, test 데이터 분리
X_train, X_test, y_train, y_test = train_test_split(wine_data,    # 데이터 feature 
                                                    wine_label,   # 데이터 label  클래스 3개
                                                    test_size=0.2,# 전체 개 20%만 테스트 데이터로 사용하겠다.
                                                    random_state=7)
print('X_train 개수: ', len(X_train), ', X_test 개수: ', len(X_test))

X_train 개수:  142 , X_test 개수:  36


이미지 데이터 178개에서 142개의 훈련데이터, 36개의 테스트데이터

# 다양한 모델로 학습시켜보기

In [264]:
from sklearn.tree import DecisionTreeClassifier # 의사결정트리 패키지
from sklearn.ensemble import RandomForestClassifier # RandomForest 모델
from sklearn import svm # svm모델
from sklearn.linear_model import SGDClassifier #SGD모델
from sklearn.linear_model import LogisticRegression #로지스틱 회귀 모델 
from sklearn.metrics import classification_report  # 정확도(precision,recall,f1-score)

### Decision Tree 사용해 보기

In [265]:
decision_tree = DecisionTreeClassifier(random_state=32)
decision_tree.fit(X_train, y_train)
y_pred = decision_tree.predict(X_test)
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00         7
           1       0.89      1.00      0.94        17
           2       1.00      0.83      0.91        12

    accuracy                           0.94        36
   macro avg       0.96      0.94      0.95        36
weighted avg       0.95      0.94      0.94        36



In [266]:
from sklearn.metrics import confusion_matrix

confusion_matrix(y_test, y_pred)

array([[ 7,  0,  0],
       [ 0, 17,  0],
       [ 0,  2, 10]])

### Random Forest 사용해 보기

In [267]:
random_forest = RandomForestClassifier(random_state=32)
random_forest.fit(X_train, y_train)
y_pred = random_forest.predict(X_test)
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00         7
           1       1.00      1.00      1.00        17
           2       1.00      1.00      1.00        12

    accuracy                           1.00        36
   macro avg       1.00      1.00      1.00        36
weighted avg       1.00      1.00      1.00        36



In [268]:
from sklearn.metrics import confusion_matrix

confusion_matrix(y_test, y_pred)

array([[ 7,  0,  0],
       [ 0, 17,  0],
       [ 0,  0, 12]])

### SVM 사용해 보기

In [269]:
svm_model = svm.SVC() # SVC모델 가져오기
svm_model.fit(X_train, y_train) # 모델 훈련 시키기
y_pred = svm_model.predict(X_test) # 모델 테스트하기
print(classification_report(y_test, y_pred)) # 정확도 확인

              precision    recall  f1-score   support

           0       0.86      0.86      0.86         7
           1       0.58      0.88      0.70        17
           2       0.33      0.08      0.13        12

    accuracy                           0.61        36
   macro avg       0.59      0.61      0.56        36
weighted avg       0.55      0.61      0.54        36



In [270]:
from sklearn.metrics import confusion_matrix

confusion_matrix(y_test, y_pred)

array([[ 6,  0,  1],
       [ 1, 15,  1],
       [ 0, 11,  1]])

### SGD Classifier 사용해 보기

In [271]:
sgd_model = SGDClassifier() #sgd_model 가져오기



sgd_model.fit(X_train, y_train) # 모델 훈련 시키기
y_pred = sgd_model.predict(X_test) # 모델 테스트기기
y_pred
print(classification_report(y_test, y_pred)) # 정확도 확인

              precision    recall  f1-score   support

           0       1.00      0.71      0.83         7
           1       0.75      0.71      0.73        17
           2       0.53      0.67      0.59        12

    accuracy                           0.69        36
   macro avg       0.76      0.70      0.72        36
weighted avg       0.73      0.69      0.70        36



### 경고
UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.   _warn_prf(average, modifier, msg_start, len(result))

In [272]:
from sklearn.metrics import confusion_matrix

confusion_matrix(y_test, y_pred)

array([[ 5,  0,  2],
       [ 0, 12,  5],
       [ 0,  4,  8]])

### Logistic Regression 사용해 보기

In [273]:
logistic_model = LogisticRegression(random_state = 32, max_iter = 4000)#logistic_model 가져오기
#max_iter 해를 찾아가는 데 있어서 반복횟수를 제한한다. 무한루프방지

logistic_model.fit(X_train, y_train) # 모델 훈련 시키기
y_pred = logistic_model.predict(X_test) # 모델 테스트기기
print(classification_report(y_test, y_pred)) # 정확도 확인

              precision    recall  f1-score   support

           0       1.00      1.00      1.00         7
           1       0.94      1.00      0.97        17
           2       1.00      0.92      0.96        12

    accuracy                           0.97        36
   macro avg       0.98      0.97      0.98        36
weighted avg       0.97      0.97      0.97        36



max_iter 해를 찾아가는 데 있어서 반복횟수를 제한한다. 무한루프방지  
max_iter인자를 추가 하지 않았을때 경고 뜸  
https://blog.naver.com/gdpresent/221703566189 참조

In [274]:
from sklearn.metrics import confusion_matrix

confusion_matrix(y_test, y_pred)

array([[ 7,  0,  0],
       [ 0, 17,  0],
       [ 0,  1, 11]])

와인은 178개 이미지, 13개 특징, 3개의 클래스   

<div style="font-size:115%; background-color:#eafbee;" >
각 모델에 대한 오차행렬을 봤을때 SVM와 SGD 빼고 3개의 모델의 성능지표 비중이 거의 TP와 TN로 구성되어 있다.즉 손글씨를 판단했을때 처럼 F1 socre를 성능지표로 평가 하기 좋을것같다. SVM와 SGD를 오차행렬로 깊숙하게 들어갈수도 있으나, "𝐹𝑃와 𝐹𝑁 값이 꽤 있어 정밀도와, 재현율의 값이 낮구나"라고 이해하겠다.(다중 오차행렬 이해하기 어렵다...)    
    
모델중 Logistic Regression,Random Forest,Decision Tree 는 95%이상,Logistic Regression,Random Forest 는 저조한 성능지표를 보여준다.특히 Random Forest의 각각의 precision,recall,f1-score, accuracy가 1로 가장 높은 값을 가진다. 그래서 모델로써는 Random Forest가 적절하다고 생각된다.    

</div>

# 유방암

In [242]:
# 필요한 모듈 import하기
from sklearn.datasets import load_breast_cancer # 사이킷런 토이데이터셋(유방암)
from sklearn.model_selection import train_test_split # 분리를 위한 패키지
from sklearn.metrics import classification_report # 정밀도, 재현율, F1 score 제공하는 패키지

#데이터 준비
breast_cancer = load_breast_cancer()
breast_cancer.keys() #breast_cancer.key() 어떤 정보들이 담겼을지, keys() 라는 메서드

dict_keys(['data', 'target', 'target_names', 'DESCR', 'feature_names', 'filename'])

In [243]:
#Feature Data 지정하기
breast_cancer_data = breast_cancer.data
breast_cancer_data.shape # 569개의 데이터와 30개 특징

(569, 30)

In [244]:
#Label Data 지정하기
breast_cancer_label = breast_cancer.target
#Target Names 출력해 보기
breast_cancer.target_names # 라벨의 이름을 확인한다.

array(['malignant', 'benign'], dtype='<U9')

In [245]:
#데이터 Describe 해 보기
print(breast_cancer.DESCR)

.. _breast_cancer_dataset:

Breast cancer wisconsin (diagnostic) dataset
--------------------------------------------

**Data Set Characteristics:**

    :Number of Instances: 569

    :Number of Attributes: 30 numeric, predictive attributes and the class

    :Attribute Information:
        - radius (mean of distances from center to points on the perimeter)
        - texture (standard deviation of gray-scale values)
        - perimeter
        - area
        - smoothness (local variation in radius lengths)
        - compactness (perimeter^2 / area - 1.0)
        - concavity (severity of concave portions of the contour)
        - concave points (number of concave portions of the contour)
        - symmetry 
        - fractal dimension ("coastline approximation" - 1)

        The mean, standard error, and "worst" or largest (mean of the three
        largest values) of these features were computed for each image,
        resulting in 30 features.  For instance, field 3 is Mean Radius, f

In [246]:
#train, test 데이터 분리

X_train, X_test, y_train, y_test = train_test_split(breast_cancer_data,    # 데이터 feature 
                                                    breast_cancer_label,   # 데이터 label   ['malignant', 'benign'] 2개
                                                    test_size=0.2,# 전체 중 20%만 테스트 데이터로 사용하겠다.
                                                    random_state=7)
print('X_train 개수: ', len(X_train), ', X_test 개수: ', len(X_test))

X_train 개수:  455 , X_test 개수:  114


이미지 데이터 569개에서 455개의 훈련데이터, 114개의 테스트데이터

# 다양한 모델로 학습시켜보기

In [247]:
# (1) 필요한 모듈 import
from sklearn.tree import DecisionTreeClassifier # 의사결정트리 패키지
from sklearn.ensemble import RandomForestClassifier # RandomForest 모델
from sklearn import svm # svm모델
from sklearn.linear_model import SGDClassifier #SGD모델
from sklearn.linear_model import LogisticRegression #로지스틱 회귀 모델 
from sklearn.metrics import classification_report  # 정확도(precision,recall,f1-score)

### Decision Tree 사용해 보기

In [248]:
# 모델 학습 및 예측
decision_tree = DecisionTreeClassifier(random_state=32)
decision_tree.fit(X_train, y_train)
y_pred = decision_tree.predict(X_test)
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.92      0.82      0.87        40
           1       0.91      0.96      0.93        74

    accuracy                           0.91       114
   macro avg       0.91      0.89      0.90       114
weighted avg       0.91      0.91      0.91       114



In [249]:
from sklearn.metrics import confusion_matrix

confusion_matrix(y_test, y_pred)

array([[33,  7],
       [ 3, 71]])

### Random Forest 사용해 보기

In [250]:
random_forest = RandomForestClassifier(random_state=32)
random_forest.fit(X_train, y_train)
y_pred = random_forest.predict(X_test)
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        40
           1       1.00      1.00      1.00        74

    accuracy                           1.00       114
   macro avg       1.00      1.00      1.00       114
weighted avg       1.00      1.00      1.00       114



In [251]:
from sklearn.metrics import confusion_matrix

confusion_matrix(y_test, y_pred)

array([[40,  0],
       [ 0, 74]])

### SVM 사용해 보기 사용해 보기

In [252]:
svm_model = svm.SVC() # SVC모델 가져오기
svm_model.fit(X_train, y_train) # 모델 훈련 시키기
y_pred = svm_model.predict(X_test) # 모델 테스트하기
print(classification_report(y_test, y_pred)) # 정확도 확인

              precision    recall  f1-score   support

           0       1.00      0.72      0.84        40
           1       0.87      1.00      0.93        74

    accuracy                           0.90       114
   macro avg       0.94      0.86      0.89       114
weighted avg       0.92      0.90      0.90       114



In [253]:
from sklearn.metrics import confusion_matrix

confusion_matrix(y_test, y_pred)

array([[29, 11],
       [ 0, 74]])

### SGD Classifier 사용해 보기<br>

In [254]:
sgd_model = SGDClassifier() #sgd_model 가져오기
sgd_model.fit(X_train, y_train) # 모델 훈련 시키기
y_pred = sgd_model.predict(X_test) # 모델 테스트기기
print(classification_report(y_test, y_pred)) # 정확도 확인

              precision    recall  f1-score   support

           0       0.91      0.78      0.84        40
           1       0.89      0.96      0.92        74

    accuracy                           0.89       114
   macro avg       0.90      0.87      0.88       114
weighted avg       0.90      0.89      0.89       114



In [255]:
from sklearn.metrics import confusion_matrix

confusion_matrix(y_test, y_pred)

array([[31,  9],
       [ 3, 71]])

### Logistic Regression 사용해 보기

In [256]:
logistic_model = LogisticRegression(random_state = 32, max_iter = 4000)#logistic_model 가져오기
#max_iter 해를 찾아가는 데 있어서 반복횟수를 제한한다. 무한루프방지

logistic_model.fit(X_train, y_train) # 모델 훈련 시키기
y_pred = logistic_model.predict(X_test) # 모델 테스트기기
print(classification_report(y_test, y_pred)) # 정확도 확인

              precision    recall  f1-score   support

           0       1.00      0.85      0.92        40
           1       0.93      1.00      0.96        74

    accuracy                           0.95       114
   macro avg       0.96      0.93      0.94       114
weighted avg       0.95      0.95      0.95       114



max_iter 해를 찾아가는 데 있어서 반복횟수를 제한한다. 무한루프방지  
max_iter인자를 추가 하지 않았을때 경고 뜸  
https://blog.naver.com/gdpresent/221703566189 참조

In [257]:
from sklearn.metrics import confusion_matrix

confusion_matrix(y_test, y_pred)

array([[34,  6],
       [ 0, 74]])

malignant, benign 설명  
https://www.doctorsnews.co.kr/news/articleView.html?idxno=20924&sc_word=&sc_word2=

<div style="font-size:115%; background-color:#eafbee;" >
암을 판단하는 경우 정밀도(TP,FP)보다는 재현율(TP,FN)에 더욱 집중적으로 봐야된다.왜냐하면 양성인사람에게 음성예측할 경우 치명적이기 때문이다.  

유방암 class는 2가지가 있는데 'malignant'와 'benign'이다. malignant는 암이라고 하고,benign는 전이하지 않는 종양이라고 이해하겠다.  
malignant = class0,  benign=class1  
class0은 종양을 암을 판단하는기준, class1는 종양을 암이 아닌것으로 판단하는 기준이기 때문에 class0은 recall 지표가 중요하고, class1은 precision 중요하다.    
  
유방암 모델을 선별함에 있어서 성능지표를 살펴보았는데 Random Forest모델이  precision, recall, f1-score 1,accracy가 1로 나왔다. 다른 모델에 월등했다. 즉 유방암판단 모델로써는 Random Forest 적절하다고 생각된다.
    

</div>
그런데 SGD Classifier 모델의 성능 지표값은 계속 변경된다. 왜 그런지 찾아보았지만 결론을 도출하지 못하였다. 추후 노드를 진행하면서 천천히 이해하기로 했다.


 

1.오차행렬로 성능지표를 판단하는구나?   
2 오차행렬 예측 결과(TP, TN, FP, FF) 이해가 되지가 않는구나...  
3.오차행렬로 정밀도,재현율,정확도,fi score 파악하기  
4.오차행렬이 2x2행령으로만 표현되는것이 아니구나( 다중오차행렬 )   
5.모델에 하나에 대한 성능지표값이 있는것이 아니구나(모델의 클래스/라벨에 대한 각각의 정밀도,재현율,정확도,f1 score 존재)   
   

### 오차 행렬에서 TN,TP,FP,FN 개념이 이해하는데 어려움이 있었다. (말싸움하는 것 같다) 또한 노드에는 이진 오차 행렬을 설명해주고 있어서 '모든 모델은 2진 오차 행렬만 갖겠구나?' 하는 착각에 빠졌는데, 다행히 슬랙에서 다른 교육생분의 다중 오차 행렬에 대한 사이트를 제공해주셔서 이해하는 데 도움이 됐다. 모델 5개에 대한 훈련과 성능지표를 확인하면서 왜 이 모델은 이렇게 나오고, 저렇게 나오지? 라는 의문이 들었는데 이것 추후 진도를 나간다면 답을 찾을 수 있을 것이라고 생각해야겠다 